### ISSUE TITLE
[unstable] Complex Hybrid. + Interaction Tail Fit Problem

### ISSUE THREAD

**Issue by Wentzell** [TRIQS/cthyb#75]
Minjae ran into a problem running a calculation with complex Hybrd. and Interaction (see appended image). It seems that it tries to call a routine intended for fitting a real-valued gf with a complex one.

The following change in the TRIQS installation used fixed the problem 

Changing line 139 in $TRIQS_PATH/lib/python2.7/site-packages/pytriqs/gf/tools.py to

    for name, sig in Sigma_iw: sig.fit_tail(fit_known_moments[name], fit_max_moment, -fit_max_n, -fit_min_n, fit_min_n, fit_max_n)

![image](https://user-images.githubusercontent.com/9369112/30428278-49070976-9953-11e7-878a-c36dca24cb5e.png)

**Comment by gkraberger** [comment]
Hi,

yes, the call for complex fitting and real fitting is different. For complex calculations, I never let the solver fit but do it afterwards using fit_tail by myself.

However, complex fitting is broken already in the TRIQS library (I think this is still the case in unstable @mferrero ?) and should not be used as-is in the moment.
@mferrero I guess I will produce a pull request for this... by now many people need this!

For your solution there should be additional `-1`s if you want a symmetric fit window (due to the Matsubara frequency w_0 = - w_-1)
```
for name, sig in Sigma_iw: sig.fit_tail(fit_known_moments[name], fit_max_moment, -fit_max_n-1, -fit_min_n-1, fit_min_n, fit_max_n)
```

Best
Gernot

**Comment by Wentzell** [comment]
Hi,

Maybe one could already adjust the calls to the tail fitting in CTHyb to work also for the complex case on unstable? Your bug fixes to TRIQS should be hopefully merged quickly.

**Comment by gkraberger** [comment]
I included a change to `tools.py` in my pull request, basically with that change. I'm not sure whether that is enough to make it work with CThyb (from what you wrote above I think it is). 

That solution only allows symmetric fit windows, one might think about extending both the interface of the `tail_fit` routine in `tools.py` and of the CThyb solver to allow the more general specification of two windows. Although I'm not sure whether practical use cases ever require non-symmetric fit windows.

**Comment by Wentzell** [comment]
With your pull request all the tests of the complex version of cthyb pass?

With the small fix I posted above I am getting some severe test failures in 
```single_site_bethe```. See below.

```
Traceback (most recent call last):
  File "/home/nils.wentzell/Coding/cthyb/test/python/single_site_bethe.py", line 57, in <module>
    assert_block_gfs_are_close(Results["Sigma_iw"], S.Sigma_iw)
  File "/home/nils.wentzell/opt/triqs/lib/python2.7/site-packages/pytriqs/utility/comparison_tests.py", line 28, in assert_block_gfs_are_close
    assert_gfs_are_close(x, y, precision)
  File "/home/nils.wentzell/opt/triqs/lib/python2.7/site-packages/pytriqs/utility/comparison_tests.py", line 22, in assert_gfs_are_close
    assert_arrays_are_close(a.data, b.data, precision)
  File "/home/nils.wentzell/opt/triqs/lib/python2.7/site-packages/pytriqs/utility/comparison_tests.py", line 5, in assert_arrays_are_close
    assert  d< precision, "Arrays are different. Difference is %s.\n %s %s"%(d,a,b)
AssertionError: Arrays are different. Difference is 3201.66789002.
 [[[ 0.18229214-3.03014111j]]

 [[ 0.13873904-3.03310168j]]

 [[ 0.09505809-3.03606804j]]

 ..., 
 [[ 0.09505809+3.03606804j]]

 [[ 0.13873904+3.03310168j]]

 [[ 0.18229214+3.03014111j]]] [[[-2295.59583976+1140.75878054j]]

 [[-2297.63161429+1141.87334701j]]

 [[-2299.67336468+1142.99009356j]]

 ..., 
 [[-2299.67336468-1142.99009356j]]

 [[-2297.63161429-1141.87334701j]]

 [[-2295.59583976-1140.75878054j]]]
```

**Comment by gkraberger** [comment]
All complex CTHyb tests do pass in master (+my pull request), but not in unstable (+my pull request): there single_site_bethe fails both for complex and real solver, atomic_gf_complex fails when comparing the tails.

However in single_site_bethe, I see a different error which happens already in the fit routine (so I don't ever get to the point where the results are compared); I don't really understand what is going on there, something about the tail-object not having an `inverse` method?

File `build/cthyb/test/python/single_site_bethe.err`:

```
/temp/AG-CMS/TRIQS/complex/new_debian/unstable/lib/python2.7/dist-packages/pytriqs/gf/gf.py:618: UserWarning: g.N1 is deprecated and not generic. Use g.target_shape[0] instead
  warnings.warn("g.N1 is deprecated and not generic. Use g.target_shape[0] instead")
/temp/AG-CMS/TRIQS/complex/new_debian/unstable/lib/python2.7/dist-packages/pytriqs/gf/gf.py:624: UserWarning: g.N2 is deprecated and not generic. Use g.target_shape[1] instead
  warnings.warn("g.N2 is deprecated and not generic. Use g.target_shape[1] instead")
Traceback (most recent call last):
  File "/itp/MooseFS/AG-CMS/TRIQS/complex/new_debian/unstable/src/cthyb/test/python/single_site_bethe.py", line 44, in <module>
    S.solve(h_int=H, **p)
  File "/itp/MooseFS/AG-CMS/TRIQS/complex/new_debian/unstable/build/cthyb/python/cthyb/cthyb_solver.py", line 122, in solve
    fit_max_moment=fit_max_moment,fit_known_moments=fit_known_moments)
  File "/temp/AG-CMS/TRIQS/complex/new_debian/unstable/lib/python2.7/dist-packages/pytriqs/gf/tools.py", line 160, in tail_fit
    for name, g in G_iw: g.tail = inverse( inverse(G0_iw[name].tail) - Sigma_iw[name].tail )
  File "/temp/AG-CMS/TRIQS/complex/new_debian/unstable/lib/python2.7/dist-packages/pytriqs/gf/tools.py", line 35, in inverse
    assert hasattr(x,'inverse')
AssertionError
```

I don't get why we would get different errors - probably in my pull request something is wrong with the tail object so that it has no `inverse` method... (?).

Regarding your error: In principle, fitting the tail of a real Sigma with the complex fitting routines should work as well, so I'm a bit worried that this test blows up in your case. We should look into that.

**Comment by gkraberger** [comment]
Correction: I see these issues already in `unstable` without my pull request. Now I wonder why they don't  occur for you?

**Comment by Wentzell** [comment]
Indeed! I was compiling against my own fork of triqs which was still missing some recent commits. In fact I just opened in issue for that: https://github.com/TRIQS/triqs/issues/484

**Comment by gkraberger** [comment]
Hi Nils,

after Michel's fix of TRIQS/triqs#484, using my pull request for triqs/unstable, the `single_site_bethe` test now passes for me. I do not understand what went wrong when you got the `AssertionError: Arrays are different. Difference is 3201.66789002.` error; I cannot reproduce it.

I tried, for a quick and dirty Bethe lattice example, to perform the tail fit of a real-solver Sigma with the complex fit routine (including my pull request) and to then compare the fit parameters thus obtained with the fit parameters that the real fitting routine gives. The maximum difference is on the order of 10^-12, which I guess can still be blamed on "numerics". This means that the complex fitting routine at least passes the test of being able to handle the special real fit case, which reassures me that there is nothing horribly wrong.

**Comment by HugoStrand** [comment]
The complex fitting using hermitian symmetry has been implemented in the Triqs tail fitting routine. I am closing this issue. Please reopen if needed.
